cmd/snap-update-ns/main.go

```diff
@@ -53,12 +56,84 @@ func run() error {
 	if err := parseArgs(os.Args[1:]); err != nil {
 		return err
 	}
+
 	// There is some C code that runs before main() is started.
 	// That code always runs and sets an error condition if it fails.
 	// Here we just check for the error.
 	if err := BootstrapError(); err != nil {
+		// If there is no mount namespace to transition to let's just quit
+		// instantly without any errors as there is nothing to do anymore.
+		if err == ErrNoNS {
+			return nil
+		}
 		return err
 	}
-	// TODO: implement this
-	return fmt.Errorf("not implemented")
+	snapName := opts.Positionals.SnapName
+
+	// Lock the mount namespace so that any concurrently attempted invocations
+	// of snap-confine are synchronized and will see consistent state.
+	lock, err := mount.OpenLock(snapName)
+	if err != nil {
+		return fmt.Errorf("cannot open mount namespace lock file: %s", err)
+	}
+	if err := lock.Lock(); err != nil {
+		return fmt.Errorf("cannot lock mount namespace: %s", err)
+	}
+	defer lock.Close()
+
+	// Read the desired and current mount profiles. Note that missing files
+	// count as empty profiles so that we can gracefully handle a mount
+	// interface connection/disconnection.
+	desiredProfilePath := fmt.Sprintf("%s/snap.%s.fstab", dirs.SnapMountPolicyDir, snapName)
+	desired, err := mount.LoadProfile(desiredProfilePath)
+	if err != nil {
+		return fmt.Errorf("cannot load desired mount profile: %s", err)
+	}
+
+	currentProfilePath := fmt.Sprintf("%s/snap.%s.fstab", dirs.SnapRunNsDir, snapName)
+	current, err := mount.LoadProfile(currentProfilePath)
+	if err != nil {
+		return fmt.Errorf("cannot load current mount profile: %s", err)
+	}
+
+	// Compute the needed changes and perform each change if needed, collecting
+	// those that we managed to perform or that were performed already.
+	changesNeeded := mount.NeededChanges(current, desired)
+	var changesMade []mount.Change
+	for _, change := range changesNeeded {
+		if change.Action == mount.Keep {
+			changesMade = append(changesMade, change)
+			continue
+		}
+		// Read mount info each time as our operations may have unexpected
```

Comment:
I think it is ok to err on the safe side. The alternative is to say the we know exactly how the kernel (including bugs) performs mount and unmount operations so that we can simulate them here. I'm not sure I like that assumption.